src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php

```diff
@@ -50,21 +44,39 @@ protected function processValue($value, $isRoot = false)
                     $resolvedArguments[] = $argument;
                     continue;
                 }
-                if ('' === $key || '$' !== $key[0]) {
+
+                if ('' === $key || ('$' !== $key[0] && !class_exists($key) && !interface_exists($key))) {
```

Comment:
here we don't need the class_exists check: we should just throw when a binding is not used (as done for $named args already), this will make things context-free again